batsh.1
.TH batsh 1V v3.0 "Internet Connection, Inc." "Virtual Hosting Manual"
.SH NAME
batsh \- non interactive batch processing for very long running programs
.SH SYNOPSIS
.B #!/usr/bin/batsh < message

.I fd#0
must be attached to a seekable file. the script filename will be the first
argument to batsh (so
.B #!
will work normally)
.SH DESCRIPTION
.I batsh
is used to batch-process extremely long running programs with a per-user
concurrency.
.I batsh
scripts are designed to be run by
.I qmail-local(8)
or
.I cron(8)
in such a way that
.I batsh
jobs won't interfere with other users of the system.

.I batsh
programs perform part of the job in each run. their exit code is arranged
such that
.I qmail-local(8)
will consider most of the time that batsh exits a temporary failure and will
try again shortly.

As such,
.I batsh
programs can be restarted at any point without necessarily re-running earlier
jobs in the list.
.SS SCRIPT FORMAT
.I batsh
programs have a trivial format that is reasonably easy for humans to read, and
extremely easy for programs to read.

Each line is a separate command to the
.I batsh
processor.
.TP
.BR | exe...
pass exe to
.B /bin/sh\ -c
and record it's status code. if it returns 111 it will be started again, if
it returns 98, it will abort processing immediately and cause
.I batsh
to exit 0. exit code 97 is silently wrapped onto exit code 0. this is to
simplify some shell-handling. all other exits are passed up to whatever calls the
.I batsh
script when the status code is read.
.TP
.BR ! exe...
pass exe to
.B /bin/sh\ -c
and return it's status code immediately. the commands
.I exec
and
.I exit
are handled specially and will always terminate the script/session.
.TP
.BR ? q
.TP
.BR ?! q
q is a question; one of:
.BR $ env
,
.BR -X filespec
,
.B [
exe ],
.BR ` exe
,
.BR filespec

All unary operators supported by
.I test(1)
are supported using the
.BR -X
format. Using
.BR ` exe
will execute
.B exe
and if it exits 0, the test succeeds. Using
.B [
exe ] is syntactic sugar for executing

.B [ exe ]
ala test.

If the question-test succeeds, then process normally, otherwise suspend
processing until we see a line with a
.B ;
by itself.
.TP
.BR ;
move up a nesting level for comparisons.
.TP
.BR & exe
causes
.B exe
as soon as possible right before ending the current operation. in this way,
the schedular can be commanded to restart (if it's idle) as soon as possible.

Note that these trailing operations are installed when-seen (so they can be
conditionally installed). this usually means you want these operations near
the beginning of the fil.e
.TP
.BR + env=value
put env=value into the environment.
.TP
.BR - env
remove env from the environment.
.TP
.BR # line
ignore this line
.SS EXIT CODES
.I batsh
assumes 111 will be treated as a temporary failure and will get it called
again in the near future and that exit code 1 will generate a more permenent
failure (when the script is started incorrectly).

.I batsh
will otherwise pass return codes up from it's children, only keeping 127,
97, 98, and 111 for itself.

Exit code 127 is used by the
system(3)
library call if /bin/sh cannot be started.

.SS ENVIRONMENT
.I batsh
uses the
.B BATSH_CONCURRENCY
environment variable to override it's own concurrency defaults of 10
concurrent batsh programs running per-user. It is
.B not
recommended that users raise this value as it will
.B not
make
.I batsh
programs run faster or complete faster.

.B BATSH_MAXDEPTH
can specify an upper-bound limit on how many times
.I batsh
will allow itself to be nested.

.B BATSH_LEVEL
is used to determine how many times it has currently been nested.

.I batsh
also requires that
.B HOME
be set up correctly on startup.


.SS DIAGNOSTICS
.I batsh
can be traced by setting the environment variable
.B BATSH_TRACE
which will cause lots of information to be spit out on standard error
about what it's doing.

.SH FILES
~/.qmail.mproc.concurrency-*
~/.qmail.qtmp.batsh.*.*.*
~/.qmail.jlog.*

Note these can be changed at compile-time.

.SH SEE ALSO
test(1), system(3), qmail-local(8), qmail-command(8), cron(8)


CHANGELOG
2006-03-01 17:12  geocar

	* unix/batsh.c: add the recipient address to journal name (makes
	  things faster)

2006-01-27 08:46  geocar

	* unix/batsh.c: twophase hack

2006-01-23 14:21  geocar

	* unix/batsh.c: yay, fix depot-doc bug, add big lock around batsh,
	  stop dspam nonsense

2005-04-04 05:39  geocar

	* doc/batsh.1: fix typo

2005-03-03 07:40  geocar

	* unix/batsh.c: move concurrency check; set close-on-exec flag;
	  this allows the concurrency check to actually _work_ on solaris

2005-03-03 07:40  geocar

	* common/lock.c: use lockf on solaris; support for lockf

2004-12-14 11:20  geocar

	* unix/batsh.c: absolute paths for dspam, batsh increases
	  concurrency

2004-09-27 14:47  geocar

	* unix/batsh.c, doc/batsh.1: support for user-defined closures,
	  actually cleaning up better if chdir'd, getting ARG_MAX from
	  compiler-limits.

2004-09-27 14:02  geocar

	* unix/batsh.c: support commands in "!" mode that begin with "exec"
	  or "exit" :)

2004-09-24 14:47  geocar

	* unix/batsh.c, doc/batsh.1: sf-report now calls dspam cron
	  (finally), batsh has some bugfixes with regards to exec and exit
	  handling.

2004-09-23 20:20  geocar

	* unix/batsh.c: batch treats exec and exit commands specially for
	  "!" commands. this way, svfilter doesn't need crazy reordering
	  rules...

2004-07-26 18:23  geocar

	* unix/batsh.c: bugfixes for when qmail is going "really really"
	  fast, rewind more often, and line fix

2004-06-16 00:15  geocar

	* unix/batsh.c: suexec2 program add (possibly unnecessary, here's
	  to hoping!) and support for system-wide default concurrencies...

2004-06-11 17:51  geocar

	* unix/batsh.c: lowered concurrency down to 3

2004-06-11 16:19  geocar

	* unix/batsh.c: close fd1 and 2 when forking so we can disconnect
	  from qmail-local

2004-06-09 20:08  geocar

	* unix/batsh.c, common/lock.c, doc/batsh.1: make the locking code
	  sane, add tracing code to batsh, error reporting to maildir,
	  shorter journal names, overflow fixes to batsh, sanity fixes.
	  batsh now works properly. smtpd now puts timeofday in received
	  header to get a better picture on concurrency. svfilter deliver
	  script now works in sandbox, and better error reporting for
	  sf-check.

2004-06-09 15:46  geocar

	* unix/batsh.c: rename deliver.batsh to svfilter.batsh

2004-06-08 22:59  geocar

	* unix/batsh.c, doc/batsh.1: batsh supports multiple invocations
	  now, some better stat-tests to accomodate test(1) style
	  comparisons, and updates to the manual page.	new tool: maildir
	  does qmail-style maildir delivery and added first version of
	  batsh delivery script

2004-06-07 22:01  geocar

	* unix/batsh.c: added loop-protection, automatic
	  refrain-from-detach for mproc, some mproc queue-pulls, and some
	  world-writable protection.

2004-06-02 15:15  geocar

	* doc/batsh.1: lstat fixes, boot fixes

2004-06-01 13:52  geocar

	* doc/batsh.1: adding some documentation

2004-05-31 15:45  geocar

	* unix/batsh.c, common/lock.c: moved locking, also finished batsh
	  and added to deploy scripts

2004-05-28 22:29  patrick

	* unix/batsh.c: Added batsh



batsh.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <errno.h>
#include <ctype.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/file.h>
#include <time.h>

#include <limits.h>

static int trace = 0;
static char *home = 0;
static int twophase = -1;

#ifndef ARG_MAX
#define ARG_MAX	1024
#endif

#ifndef BATSH_DEFAULT_CONCURRENCY
#define BATSH_DEFAULT_CONCURRENCY 25
#endif

#ifndef BATSH_DEFAULT_MAXDEPTH
#define BATSH_DEFAULT_MAXDEPTH 10
#endif

#ifndef JOURNAL_STUB
#define JOURNAL_STUB ".qmail.jlog"
#endif

#ifndef CONCURRENCY_STUB
#define CONCURRENCY_STUB ".qmail.mproc"
#endif

#ifndef TEMPFILE_STUB
#define TEMPFILE_STUB	".qmail.qtmp"
#endif


static int sbaccess(int u, int g, struct stat *s, int m)
{
	if (s->st_uid == u) {
		if ((m << 6) & s->st_mode) return 1;
	}
	if (s->st_gid == g) {
		if ((m << 3) & s->st_mode) return 1;
	}
	if (m & s->st_mode) return 1;
	return 0;
}
static int confvar(const char *s)
{
	struct stat sb;
	const char *f;
	int st;

	while (isspace((int)*s)) s++;
	if (*s == '!') return ! confvar(s+1);

	if (*s == '$') return getenv(s+1) != NULL;
	if (s[0] == '-' && (s[1] == 'l' || s[1] == 'L')) {
		f = s+2;
		while (isspace((int)*f)) f++;
		if (lstat(f, &sb) == -1) return 0;
		if (S_ISLNK(sb.st_mode)) return 1;
		return 0;
	}
	if (*s == '[' || *s == '`') {
		/* rewind */
		if (lseek(0, 0, SEEK_SET) != (off_t)0) {
			if (trace) fprintf(stderr, "lseek SEEK_SET(%s)\n", strerror(errno));
			exit(111);
		}

		if (*s == '`') s++;
		while (isspace((int)*s)) s++;
		st = system(s);
		if (WIFEXITED(st) && WEXITSTATUS(st) == 0) return 1;
		if (!WIFEXITED(st) || WEXITSTATUS(st) == 127) {
			if (trace) fprintf(stderr, "%s = system(%s)\n", s, strerror(errno));
			exit(111);
		}
		return 0;
	}
	if (*s == '/' || *s == '.' || *s == '-') {
		if (*s == '-') {
			f = s+2;
			while (isspace((int)*f)) f++;
			if (!*f) return 0;
		} else {
			f = s;
		}

		if (stat(f, &sb) == -1) return 0;
		if (*s != '-' || s[1] == 'e') return 1;

		switch ((s[1])) {
		case 'r': return sbaccess(geteuid(), getegid(), &sb, 04);
		case 'w': return sbaccess(geteuid(), getegid(), &sb, 02);
		case 'x': return sbaccess(geteuid(), getegid(), &sb, 01);
		case 'o': return geteuid() == sb.st_uid;

		case 'R': return sbaccess(getuid(), getgid(), &sb, 04);
		case 'W': return sbaccess(getuid(), getgid(), &sb, 02);
		case 'X': return sbaccess(getuid(), getgid(), &sb, 01);
		case 'O': return getuid() == sb.st_uid;

		case 'z': return sb.st_size == 0;
		case 's': return sb.st_size != 0;

		case 'f': return S_ISREG(sb.st_mode);
		case 'd': return S_ISDIR(sb.st_mode);
		case 'p': return S_ISFIFO(sb.st_mode);
		case 'S': return S_ISSOCK(sb.st_mode);
		case 'b': return S_ISBLK(sb.st_mode);
		case 'c': return S_ISCHR(sb.st_mode);
		case 't': return 0;

		case 'u': return sb.st_mode & S_ISUID;
		case 'g': return sb.st_mode & S_ISGID;
		case 'k': return sb.st_mode & S_ISVTX;

		case 'T': return 1; /* sure */
		case 'B': return 1; /* sure */

		};
		/* otherwise, not true */
		return 0;
	}

	return 0;
}

static int last_wd = -1;
static int mcc = BATSH_DEFAULT_CONCURRENCY;

void dolock(int fd);
int trylock(int fd);

static void build_journal(char *jobs, int scriptfd);
static int cwrite(int fd, void *buf, size_t len);
static int cread(int fd, void *buf, size_t len);
static off_t read_record(int jh, unsigned char *st, char **cmd);
static void run_child(int jh, char *cmd, char *jobs);
static int build_jobs(FILE *fp, int th);
static void setup_build_jobs(FILE *fp, char *jobs);
static void do_jobs(int jh, char *jobs, int scriptfd);

static int cwrite(int fd, void *buf, size_t len)
{
	int r, x;
	for (x = 0; x < len; x++) {
		do {
			r = write(fd, (void *)(buf+x), len-x);
		} while (r == -1 && errno == EINTR);
		if (r < 1) return 0;
		x += r;
	}
	return 1;
}

static int cread(int fd, void *buf, size_t len)
{
	int r, x;
	for (x = 0; x < len; x++) {
		do {
			r = read(fd, (void *)(buf+x), len-x);
		} while (r == -1 && errno == EINTR);
		if (r == 0) return 0; 
		x += r;
	}
	return 1;
}

static off_t read_record(int jh, unsigned char *st, char **cmd)
{
	unsigned int len;
	off_t o;

	if (!cread(jh, &len, sizeof(unsigned int))) return (off_t)0;
	o = lseek(jh, 0, SEEK_CUR);
	if (o < 1) {
		/* impossible */
		return (off_t)-1;
	}

	if (!cread(jh, st, sizeof(unsigned char))) return (off_t)-1;
	*cmd = (char *)malloc(len+1);
	if (!*cmd) {
		if (trace) fprintf(stderr, "malloc(%s)\n", strerror(errno));
		exit(111);
	}
	if (!cread(jh, *cmd, len+1)) return (off_t)-1;
	cmd[0][len] = 0;

	return o;
}

/* schedule delivery as soon as is possible */
static char *closures = 0;
static unsigned int closures_len = 0;
static void add_closure(char *buf)
{
	unsigned int l;

	if (strcmp(buf, "&") == 0) {
		/* kill closures */
		free(closures);
		closures = 0;
		closures_len = 0;
		return;
	}

	l = strlen(buf)+1;
	closures = (char *)realloc(closures, closures_len + (l+1));
	if (!closures) {
		if (trace) fprintf(stderr, "realloc(%s)\n", strerror(errno));
		exit(111);
	}
	memcpy(closures+closures_len, buf, l);
	closures_len += l;
}
static void run_closures(void)
{
	unsigned int i, l;
	char *p;
	int st;

	for (i = 0, p = closures; i < closures_len;) {
		l = strlen(p)+1;
		st = system(p);
		if (WIFEXITED(st)) {
			if (trace) fprintf(stderr, "& %s = exited (%d)\n", p, WEXITSTATUS(st));
			if (WEXITSTATUS(st) != 0) {
				break;
			}
		} else {
			if (trace) fprintf(stderr, "& %s = killed\n", p);
			break;
		}

		p += l; i+= l;
	}
}

static void get_concurrency(void)
{
	if (mcc > 1) {
		static char mc_buffer[128];
		int i, fd, cx;

		for (i = cx = 0; !cx && i < mcc; i++) {
			sprintf(mc_buffer, 
					CONCURRENCY_STUB
					".concurrency-%d", i);
			fd = open(mc_buffer, O_RDWR|O_CREAT, 0600);
			if (fd > -1) {
				if (trylock(fd) == 1) {
					(void)fcntl(fd, F_SETFD, 0);
					cx = 1;
				} else {
					(void)close(fd);
				}
			}
		}
		if (!cx) {
			/* couldn't obtain concurrency */
			if (trace) fprintf(stderr, "no concurrency\n");
			exit(111);
		}
		/* fall through; got concurrency handle */
	}
}

static void run_child(int jh, char *cmd, char *jobs)
{
	char *p;
	int st, eflag;

	puts(cmd);
	/* return to working directory */
	if (fchdir(last_wd) == -1) {
		if (trace) fprintf(stderr, "fchdir(%s)\n", strerror(errno));
		run_closures();
		exit(111);
	}

	/* rewind */
	if (lseek(0, 0, SEEK_SET) != (off_t)0) {
		if (trace) fprintf(stderr, "lseek SEEK_SET(%s)\n", strerror(errno));
		exit(111);
	}

	eflag = 0;
	if (*cmd == '!') {
		if (!strncmp(cmd+1, "exit", 4)) {
			if (isspace(((unsigned int)cmd[5]))) {
				p = cmd+5;
				while (isspace(((unsigned int)*p))) p++;
				if (trace) fprintf(stderr, "exit(%d)\n", atoi(p));
				run_closures();
				if (chdir(home) != -1)
					(void) unlink(jobs); /* delete the job */
				if (trace) fprintf(stderr, "unlink(%s)\n", jobs);
				exit(atoi(p));
			}

		} else if (!strncmp(cmd+1, "exec", 4)) {
			if (isspace(((unsigned int)cmd[5]))) {
				if (trace) fprintf(stderr, "execute-flag on\n");
				eflag = 1;
			}
		}

		st = system(cmd+1);
		if (st == -1) {
			if (trace) fprintf(stderr, "%s = system(%s)\n", cmd, strerror(errno));
			run_closures();
			exit(111);
		}
		if (trace) fprintf(stderr, "returned with st: %x\n", st);

	} else if (*cmd == '|') {
		pid_t p;
		switch ((p = fork())) {
		case -1:
			if (trace) fprintf(stderr, "fork(%s)\n", strerror(errno));
			exit(111);
		case 0:
			/* child */
			if (!trace) {
				(void) close(1);
				(void) close(2);
			}
			get_concurrency();
			st = system(cmd+1);
			if (st == -1) {
				if (trace)
					fprintf(stderr, "%s = system(%s)\n", cmd, strerror(errno));
				exit(111);
			}
			break;
		default:
			/* parent */
			if (trace) fprintf(stderr, "batsh/fork()\n");
			exit(111);
		};
	} else {
		if (trace) fprintf(stderr, "%s = unknown tooling\n", cmd);
		exit(111);
	}

	/* check status */
	if (WIFEXITED(st)) {
		unsigned char sq[1];
		sq[0] = WEXITSTATUS(st);

		if (trace) {
			if (sq[0] == 127) {
				fprintf(stderr, "%s = system failed\n", cmd);
			} else {
				fprintf(stderr, "%s = %d\n", cmd, (int)sq[0]);
			}
		}

		if (sq[0] == 97) sq[0] = 0; /* hack- continue processing */

		if (sq[0] == 127 || sq[0] == 111) {
			run_closures();
			exit(111);
		}
		if (!cwrite(jh, sq, 1)) {
			if (trace) fprintf(stderr, "Failed to write to journal: %s\n", strerror(errno));
			run_closures();
			exit(111);
		}
		if (*cmd == '!') {
			if (sq[0] == 0 && !eflag) {
				if (trace) fprintf(stderr, "!return\n");
				return; /* only path for return! */
			}
			if (chdir(home) != -1)
				(void) unlink(jobs); /* delete the job */
		}
		run_closures();
		if (sq[0] == 98) sq[0] = 0; /* hack2 */
		exit(sq[0]);
	} else {
		/* other failure */
		run_closures();
		if (trace) fprintf(stderr, "%s = killed\n", cmd);
		exit(111);
	}
}

static void setup_build_jobs(FILE *sc, char *jobs)
{
	/* must be bigger than TEMPFILE_STUB+64-or-so characters */
	static char line[1024];
	int th, i;

	i = 0;
	do {
		sprintf(line, TEMPFILE_STUB ".batsh.%lu.%lu.%u",
					(unsigned long)time(0),
					(unsigned long)getpid(),
					i++);
		th = open(line, O_CREAT|O_EXCL|O_WRONLY, 0600);
	} while (th == -1 && (errno == EINTR
#ifdef EAGAIN
				|| errno == EAGAIN
#endif
				|| errno == EEXIST));
	if (th == -1) {
		if (trace) fprintf(stderr, "open/qtmp()\n");
		exit(111);
	}

	if (!build_jobs(sc, th)) {
		close(th);
		unlink(line);
		if (trace) fprintf(stderr, "build_jobs()\n");
		exit(111);
	}

	if (close(th) == -1) {
		unlink(line);
		if (trace) fprintf(stderr, "close(%s)\n", strerror(errno));
		exit(111);
	}
	
	if (rename(line, jobs) == -1) {
		unlink(line);
		if (trace) fprintf(stderr, "rename(%s)\n", strerror(errno));
		exit(111);
	}
}

static int build_jobs(FILE *sc, int th)
{
	unsigned int len;
	static char linebuf[ARG_MAX];
	char *line;
	int i;

	while (fgets(linebuf, sizeof(linebuf) - 1, sc)) {
		for (i = 0; isspace(((int)linebuf[i])); i++);
		if (! linebuf[i]) continue;
		line = &linebuf[i];

		if (*line == '#') continue;

		len = strlen(line);
		if (line[0] == '<') {
			FILE *fp;
			/* relative to $HOME! */
			fp = fopen(line+1, "r");
			if (!fp) continue;
			if (!build_jobs(fp, th)) {
				fclose(fp);
				return 0;
			}
			fclose(fp);
		} else {
			unsigned char cs[sizeof(len)+2];
			unsigned char *qx;

			if (strchr("|!?+-;&", *line) == 0) {
				cs[sizeof(len)+1] = '|';
				qx = line;
			} else {
				cs[sizeof(len)+1] = *line;
				qx = line+1;
				len--;
			}

			memcpy(cs, &len, sizeof(len));
			cs[sizeof(len)] = 111;

			if (!cwrite(th, cs, sizeof(cs))) return 0;
			if (!cwrite(th, qx, strlen(qx))) return 0;
		}
	} 
	return 1;
}



static void build_journal(char *jobs, int scriptfd)
{
	int jh;
	struct stat s, j;


	if (fstat(scriptfd, &s) == -1) {
		if (trace) fprintf(stderr, "fstat(%s)\n", strerror(errno));
		exit(111);
	}
	if (s.st_mode & 022) {
		if (trace) fprintf(stderr, "scriptfd = world writable\n");
		exit(111); /* world writable bombout! */
	}

REOPEN:
	jh = open(jobs, O_RDWR|O_NOCTTY);
	if (jh == -1 || fstat(jh, &j) == -1 || (s.st_mtime > j.st_mtime))  {
		FILE *fp;
		int xfd;

		if (jh != -1) {
			dolock(jh);
		}

		if (trace) fprintf(stderr, "mtime(%lu) <=> mtime(%lu)\n",
					(unsigned long)s.st_mtime,
					(unsigned long)j.st_mtime);
		xfd = dup(scriptfd);
		if (xfd == -1) {
			if (trace) fprintf(stderr, "dup(%s)\n", strerror(errno));
			exit(111);
		}

		fp = fdopen(xfd, "r");

		if (!fp) {
			if (trace) fprintf(stderr, "fdopen(%s)\n", strerror(errno));
			exit(111);
		}
		setup_build_jobs(fp, jobs);

		fclose(fp);
		(void) close(jh);
		goto REOPEN;
	}

	if (fstat(jh, &j) == -1) {
		if (trace) fprintf(stderr, "fstat(%s)\n", strerror(errno));
		exit(111);
	}
	if (j.st_mode & 022) {
		if (trace) fprintf(stderr, "jobsfd = world writable\n");
		exit(111); /* world writable bombout! */
	}

	if (trylock(jh) == 1) {
		(void)close(twophase);
		twophase = -1;

		do_jobs(jh, jobs, scriptfd);
		if (trace) fprintf(stderr, "do_jobs returned?\n");
	} else {
		if (trace) fprintf(stderr, "journal already locked\n");
	}
	exit(111);
}

extern char **environ;
/* because unsetenv is unportable */
static void remove_from_env(const char *s)
{
	int i, j;

	j = strlen(s);
	for (i = 0; environ[i]; i++){ 
		if (strncmp(environ[i], s, j) == 0
			&& environ[i][j] == '=') {
			for (j = i+1; environ[j]; j++) {
				environ[j-1] = environ[j];
			}
			environ[j-1] = 0;
			return;
		}
	}
}

static void do_jobs(int jh, char *jobs, int scriptfd) 
{
	unsigned char st;
	int skip = 0;
	off_t r, p;

	char *cmd = 0;
	while ((r = read_record(jh, &st, &cmd)) > 0) {
		if (st == 0) {
			free(cmd);
			continue;

		} else if (st == 111) {
			if (*cmd == ';') {
				free(cmd);
				if (skip > 0) skip--;
				continue;

			} else if (skip) {
				if (*cmd == '?') skip++;
				free(cmd);
				continue;
			}
			if (*cmd == '+') {
				if (putenv(cmd+1) == -1) {
					if (trace) fprintf(stderr, "putenv(%s)\n", strerror(errno));
					exit(111);
				}
				continue;

			} else if (*cmd == '-') {
				remove_from_env(cmd+1);
				free(cmd);
				continue;

			} else if (*cmd == '?') {
				if (confvar(cmd+1) != 0) {
					free(cmd);
					continue;
				}
				/* hrm... not in env... skipmode */
				skip++;
				free(cmd);
				continue;

			} else if (*cmd == '&') {
				add_closure(cmd+1);
				free(cmd);
				continue;
			}

			p = lseek(jh, 0, SEEK_CUR);
			if (p == (off_t)-1) {
				if (trace) fprintf(stderr, "lseek SEEK_CUR(%s)\n", strerror(errno));
				exit(111);
			}
			if (lseek(jh, r, SEEK_SET) != r) {
				if (trace) fprintf(stderr, "lseek SEEK_SET(%s)\n", strerror(errno));
				exit(111);
			}
			run_child(jh, cmd, jobs);
			free(cmd);
			if (lseek(jh, p, SEEK_SET) != p) {
				if (trace) fprintf(stderr, "lseek SEEK_SET(%s)\n", strerror(errno));
				exit(111);
			}

		} else {
			if (st == 98) st = 0;
			if (trace) fprintf(stderr, "journal command exit: %d\n", st);
			if (chdir(home) != -1)
				(void) unlink(jobs);
			exit(st);
		}
	}
	if (r == (off_t)-1) {
		if (trace) fprintf(stderr, "script changing too fast\n");
		exit(111); /* eek... try again later */
	}
	/* all done! if we get here, then r == 0, and we're at the
	 * end of the file.
	 */
	if (chdir(home) != -1)
		(void) unlink(jobs);
	if (trace) fprintf(stderr, "script completion\n");
	exit(0);
}

static int capture_int(const char *f, int def)
{
	FILE *fp;
	int o = 0, c;
	fp = fopen(f, "r");
	if (!fp) return def;
	for (;;) {
		c = fgetc(fp);
		if (c == -1) break;
		if (!isdigit(c)) break;

		o *= 10;
		o += (c - '0');
	}
	(void) fclose(fp);
	if (o == 0) return def;
	return o;
}

int main(int argc, char **argv) 
{
	static char sb[64];
	char *recipient, *jobs;
	struct stat inf;
	struct stat sinf;
	char *q;
	int bl = 0;
	int bm;
	int fd;

	recipient = getenv("RECIPIENT");
	if (!recipient) {
		recipient = "";
	} else {
		for (q = recipient; *q; q++)
			if (*q == '@') *q = '=';
			else if (*q == '.') *q = ':';
			else if (*q == '/') *q = '_';
	}
	jobs = malloc(sizeof(JOURNAL_STUB) + 64 + strlen(recipient));
	if (!jobs) {
		perror("malloc");
		exit(1);
	}

	trace = (getenv("BATSH_TRACE")) ? 1 : 0;
	if (trace) setlinebuf(stderr);
	if (trace) fprintf(stderr, "batsh starting up\n");

	/* set concurrency */
	q = (char *)getenv("BATSH_CONCURRENCY");
	if (!q) mcc = capture_int("/etc/batsh.concurrency", BATSH_DEFAULT_CONCURRENCY);
	else {
		mcc = atoi(q);
		if (mcc < 1) mcc = capture_int("/etc/batsh.concurrency", BATSH_DEFAULT_CONCURRENCY);
	}

	/* depth limit */
	q = (char *)getenv("BATSH_MAXDEPTH");
	if (!q) bm = BATSH_DEFAULT_MAXDEPTH;
	else {
		bm = atoi(q);
		if (bm < 1) bm = BATSH_DEFAULT_MAXDEPTH;
	}

	/* increment batsh level (depthfinder) */
	q = (char *)getenv("BATSH_LEVEL");
	if (q) bl = atoi(q);
	if (bl < 1) {
		bl = 1;
	} else if (bl > bm) {
		fprintf(stderr, "$BATSH_LEVEL greater than $BATSH_MAXDEPTH (%d > %d)\n", bl, bm);
		exit(1);
	} else {
		bl++;
	}
	sprintf(sb, "BATSH_LEVEL=%d", bl);
	putenv(sb);

	if (fstat(0, &inf) == -1) {
		if (trace) fprintf(stderr, "Can't fstat() stdin: %s\n", strerror(errno));
		exit(111);
	}
	if (lseek(0, 0, SEEK_SET) != (off_t)0
			|| argc != 2 || !S_ISREG(inf.st_mode)) {
		fprintf(stderr, "Usage: #!batsh < qmess\n");
		exit(1); 
	}

	last_wd = open(".", O_RDONLY);
	if (last_wd == -1) {
		if (trace) fprintf(stderr, "open(.): %s\n", strerror(errno));
		exit(111); /* no current working directory? */
	}

	q = (char *)getenv("HOME");
	if (!q) {
		fprintf(stderr, "$HOME is unset\n");
		exit(1);
	}
	home = strdup(q);
	if (!home) {
		if (trace) fprintf(stderr, "Can't strdup(): %s\n", strerror(errno));
		exit(111);
	}

	twophase = open(".email/.svfilter.twophase", O_CREAT|O_RDWR,0600);
	if (twophase == -1) {
		if (trace) fprintf(stderr, "Cannot open two-phase lock head: %s\n",strerror(errno));
		if (trylock(twophase) != 1) {
			if (trace) fprintf(stderr, "another twophase holder open\n");
			exit(111);
		}
	}

	fd = open(argv[1], O_RDONLY|O_NOCTTY);
	if (fd == -1) {
		if (trace) fprintf(stderr, "open(%s): %s\n", argv[1], strerror(errno));
		exit(111);
	}
	if (fstat(fd, &sinf) == -1) {
		if (trace) fprintf(stderr, "Can't fstat() script: %s\n", strerror(errno));
		exit(111);
	}

	if (chdir(home) == -1) {
		if (trace) fprintf(stderr, "Can't chdir(%s): %s\n", q, strerror(errno));
		exit(111);
	}

	sprintf(jobs, "%s.%lx:%lx.%lx:%lx;%s", JOURNAL_STUB,
			(long int)inf.st_dev,  (long int)inf.st_ino,
			(long int)sinf.st_dev, (long int)sinf.st_ino,
			recipient);
	
	build_journal(jobs, fd);

	/* should not reach */
	if (trace) fprintf(stderr, "Should not reach\n");
	exit(1);
}
